In [1]:
import pandas as pd
import requests
from sqlalchemy import create_engine
import json

# Artists
## Rage Stage (rap stage)
- Lil uzi vert
- Chance the Rapper
- Trippie Redd
- YBN Cordae
- Kanye West

## Unwind (acoustic/soft rock stage)
- Jack Johnson
- John Mayer
- The Head and the Heart
- Khrubangbin
- Rex Orange County

## The Haus (dance stage)
- Ricky Retro
- Flume
- Louis the Child
- Hippie Sabatoge
- Fisher

## Hick Stage (country stage)
- Jake Owen
- Kenny Chesney
- Thomas rhett
- Luke Bryan
- Jon Pardi

## Mane Stage (most popular)
- Justin Beiber
- Ed Sheeran
- Taylor Swift
- Lorde
- Frank Ocean


## Generate Authorization Token
Link for access token : https://developer.spotify.com/console/get-search-item/

In [83]:
token = "BQBnZ9cligHceuph084_f0DQ1n8pFBVw1RvO1unC9qYkCQWZjmIEjoJ9oUAhtBZ1E_aDy_9wGWFWptnk8YorZcLLY46m2vPjcu1Jx3nrcI1qNrkKWGE72Xj4wuKIG93Geq3JUl3-AwkGZ7Ck_PoWcvcO3lXbu4QxEEyg1ZospBGWKwj6mFMZ5XXmyPT6lsCDfO3sfQymtcN2q4AmosgkvKatzso7Rrbf-mzGpnNrjr6O9tv2KlTYfDUlHAvhDVF9Rn2xINZJOwFUSms"


## Populate the Artist Table

In [92]:
def get_artist(token,artist):
    """
    The following function, takes advantage of Spotify's search endpoint to return an artist's ID
    based off of their name. 
    Parameters
    -----------
    Token- temporary authorization token generated by spotify
    artist - name of the desired artist to get info for
    -----------
    Link for access token: https://developer.spotify.com/console/get-search-item/
    """
    artist_spotify_url = "https://api.spotify.com/v1/search"
    api_token = token #must get token from spotify to access data
    artist_headers = {"Authorization" : "Bearer %s"%api_token, "Content-Type": "application/json"} 
    artist_params = {"type": "artist", "q":artist}
    artist_request = requests.get(artist_spotify_url, params=artist_params, headers=artist_headers)
    loaded_data = json.loads(artist_request.text) #get raw data in json format
    follower_count = [] #store follower counts for search results
    artist_dict = {}# store follower count as key and artist ID and name's as value 
    for i in range(len(json.loads(artist_request.text)["artists"]["items"])):
        ID = (loaded_data["artists"]["items"][i]["id"])#artist's unique id
        name = (loaded_data["artists"]["items"][i]["name"])
        followers = (loaded_data["artists"]["items"][i]["followers"]['total'])
        popularity = (loaded_data["artists"]["items"][i]["popularity"])
        artist_dict[followers] = [ID,name,popularity,followers]
        follower_count.append(followers)
        true_artist = max(follower_count)
        true_artist_dict = artist_dict[true_artist]
        return true_artist_dict

In [93]:
#Populate artist_table using search
artists = ["Lil uzi vert", "Chance the Rapper", "Trippie Redd", "YBN Cordae", "Kanye West", "Jack Johnson", "John Mayer",
           "The Head and the Heart", "Khrubangbin", "Rex Orange County", "Ricky Retro", "Flume", "Louis the Child", 
           "Hippie Sabatoge", "Fisher", "Jake Owen", "Kenny Chesney", "Thomas rhett", "Luke Bryan", "Jon Pardi",
           "Justin Beiber", "Ed Sheeran", "Taylor Swift", "Lorde", "Frank Ocean"
          ]
#create dictionary for dataframe
artists_dict = {
    "id" : [],
    "name" : [],
    "popularity" : [],
    "followers" : []
    }
#iterate through artists to populate dictionary with appropriate data
for x in artists:
    artists_dict["name"].append(x) #pass in name from list
    artists_dict["id"].append(get_artist(token,x)[0])
    artists_dict["popularity"].append(get_artist(token,x)[2])
    artists_dict["followers"].append(get_artist(token,x)[3])

In [94]:
#assign artists to a dataframe
artists = pd.DataFrame(artists_dict)

In [95]:
artists

id                    name  popularity  followers
0   4O15NlyKLIASxsJ0PrXPfz            Lil uzi vert          95    8436114
1   1anyVhU62p31KFi8MEzkbf       Chance the Rapper          83    4770454
2   6Xgp2XMz1fhVYe7i6yNAax            Trippie Redd          88    4510836
3   0huGjMyP507tBCARyzSkrv              YBN Cordae          73    1010739
4   5K4W6rqBFWDnAN6FQUkS6x              Kanye West          90   11528211
5   3GBPw9NK25X1Wt2OUvOwY3            Jack Johnson          80    2766315
6   0hEurMDQu99nJRq8pTxO14              John Mayer          83    4491072
7   0n94vC3S9c3mb2HyNAOcjg  The Head and the Heart          71     802721
8   2mVVjNmdjXZZDvhgQWiakk             Khrubangbin          71     528117
9   7pbDxGE6nQSZVfiFdq9lOL       Rex Orange County          81    1972337
10  1fA8WmzG2eU2q13kAAjoN8             Ricky Retro          50      11377
11  6nxWCVXbOlEVRexSbLsTer                   Flume          78    1988082
12  7wg1qvie3KqDNQbAkTdbX0         Louis the Child          73     347909
13  4dM6NDYSfLcspt8GLoT5aE         Hippie Sabatoge          75     476532
14  1VJ0briNOlXRtJUAzoUJdt                  Fisher          73     386261
15  1n2pb9Tsfe4SwAjmUac6YT               Jake Owen          71    1357132
16  3grHWM9bx2E9vwJCdlRv9O           Kenny Chesney          76    2904195
17  6x2LnllRG5uGarZMsD4iO8            Thomas rhett          79    3318620
18  0BvkDsjIUla7X0k6CSWh1I              Luke Bryan          79    5185640
19  4MoAOfV4ROWofLG3a3hhBN               Jon Pardi          75     979328
20  1uNFoZAHBGtllmzznpCI3s           Justin Beiber          95   34270040
21  6eUKZXaKkcviH0Ku9w2n3V              Ed Sheeran          93   62327775
22  06HL4z0CvFAxyc27GXpf02            Taylor Swift          92   28345596
23  163tK9Wjr9P9DmM0AVK7lm                   Lorde          79    5665541
24  2h93pZq0e7k5yf4dywlkpM             Frank Ocean          85    5557798

In [98]:
artists.to_csv("artist.csv", index=False)

## Populate the Album Table

In [31]:
def add_albums(album_data,passed_dict):
    for i in range(len(album_data['items'])):
        name = album_data['items'][i]['name']
        passed_dict['name'].append(name)
        ID = album_data['items'][i]['id']
        passed_dict['album_id'].append(ID)
        total_tracks = album_data['items'][i]['total_tracks']
        passed_dict['total_tracks'].append(total_tracks)
        release_date = album_data['items'][i]['release_date']
        passed_dict['release_date'].append(release_date)
        album_type = album_data['items'][i]['album_type']
        passed_dict['album_type'].append(album_type)

In [36]:
def add_all_albums (artist_id, album_type):
    album_spotify_url = "https://api.spotify.com/v1/artists/{}/albums".format(artist_id)
    api_token = token #must get token from spotify to access data
    headers = {"Authorization" : "Bearer %s"%api_token, "Content-Type": "application/json"} 
    offset = 0
    params = {"include_groups":  album_type, "offset": str(offset)}
    request_og = requests.get(album_spotify_url, headers=headers, params=params)
    loaded_data_og = json.loads(request_og.text) #get raw data in json format
    albums_dict ={
        "album_id":[],
        "name":[],
        "total_tracks":[],
        "release_date":[],
        "album_type":[]
    }
    add_albums(loaded_data_og, albums_dict)
    while len(albums_dict["album_id"]) < loaded_data_og['total']:
        offset +=20
        params["offset"] = str(offset)
        request = requests.get(album_spotify_url, headers=headers, params=params)
        loaded_data = json.loads(request.text)
        add_albums(loaded_data, albums_dict)
    output = pd.DataFrame(albums_dict)
    output["artist_id"] = artist_id
    return output

In [40]:
artist_ids = [x for x in artists['id']]

In [42]:
album_df = pd.DataFrame()
for x in artist_ids:
    album1_df = add_all_albums(x, "single")
    album_df = album_df.append(album1_df)
    album2_df = add_all_albums(x,"album")
    album_df = album_df.append(album2_df)

album_id                               name  total_tracks  \
0   4iz9Pqwpw8HljZei6XFCGd                           That Way           1.0   
1   4uTOUYhLrGRKkfxfPKTkuy                Futsal Shuffle 2020           1.0   
2   5UL4JeebFRM3CwLfk7ZAlh    Show U Off (feat. Lil Uzi Vert)           1.0   
3   2WyjmS6PEAE2V9oh5So588    Show U Off (feat. Lil Uzi Vert)           1.0   
4   49K82qcZN4u109e6Zw8io1                  Sanguine Paradise           1.0   
..                     ...                                ...           ...   
15  3TLaFWQDhV1g39Qwd5sPAm                          Swim Good           1.0   
0   3mH6qwIy9crq0I9YQbOuDf                             Blonde          17.0   
1   392p3shh2jkxUxY2VHvlH8                     channel ORANGE          17.0   
2   6vzH59tobnxE5WXMiqvBIx                     channel ORANGE          17.0   
3   623Ef2ZEB3Njklix4PC0Rs  channel ORANGE (Explicit Version)          17.0   

   release_date album_type               artist_id  
0    2020-03-01     single  4O15NlyKLIASxsJ0PrXPfz  
1    2019-12-12     single  4O15NlyKLIASxsJ0PrXPfz  
2    2019-04-26     single  4O15NlyKLIASxsJ0PrXPfz  
3    2019-04-26     single  4O15NlyKLIASxsJ0PrXPfz  
4    2019-04-09     single  4O15NlyKLIASxsJ0PrXPfz  
..          ...        ...                     ...  
15   2011-01-01     single  2h93pZq0e7k5yf4dywlkpM  
0    2016-08-20      album  2h93pZq0e7k5yf4dywlkpM  
1    2012-01-01      album  2h93pZq0e7k5yf4dywlkpM  
2    2012-01-01      album  2h93pZq0e7k5yf4dywlkpM  
3    2012-01-01      album  2h93pZq0e7k5yf4dywlkpM  

[939 rows x 6 columns]

In [46]:
album_ids = [album_id for album_id in album_df["album_id"] ]

In [72]:
def get_album_popularity(album_id):
    album_spotify_url = "https://api.spotify.com/v1/albums/{}".format(album_id)
    api_token = token #must get token from spotify to access data
    headers = {"Authorization" : "Bearer %s"%api_token, "Content-Type": "application/json"} 
    request = requests.get(album_spotify_url, headers=headers)
    loaded_test_data = json.loads(request.text) #get raw data in json format
    return loaded_test_data['popularity']

In [87]:
for album_id in album_ids:
    album_df.loc[album_df.album_id == album_id, 'popularity'] = get_album_popularity(album_id)

In [88]:
album_df

album_id                               name  total_tracks  \
0   4iz9Pqwpw8HljZei6XFCGd                           That Way           1.0   
1   4uTOUYhLrGRKkfxfPKTkuy                Futsal Shuffle 2020           1.0   
2   5UL4JeebFRM3CwLfk7ZAlh    Show U Off (feat. Lil Uzi Vert)           1.0   
3   2WyjmS6PEAE2V9oh5So588    Show U Off (feat. Lil Uzi Vert)           1.0   
4   49K82qcZN4u109e6Zw8io1                  Sanguine Paradise           1.0   
..                     ...                                ...           ...   
15  3TLaFWQDhV1g39Qwd5sPAm                          Swim Good           1.0   
0   3mH6qwIy9crq0I9YQbOuDf                             Blonde          17.0   
1   392p3shh2jkxUxY2VHvlH8                     channel ORANGE          17.0   
2   6vzH59tobnxE5WXMiqvBIx                     channel ORANGE          17.0   
3   623Ef2ZEB3Njklix4PC0Rs  channel ORANGE (Explicit Version)          17.0   

   release_date album_type               artist_id popularity  
0    2020-03-01     single  4O15NlyKLIASxsJ0PrXPfz         76  
1    2019-12-12     single  4O15NlyKLIASxsJ0PrXPfz         73  
2    2019-04-26     single  4O15NlyKLIASxsJ0PrXPfz         33  
3    2019-04-26     single  4O15NlyKLIASxsJ0PrXPfz         38  
4    2019-04-09     single  4O15NlyKLIASxsJ0PrXPfz         71  
..          ...        ...                     ...        ...  
15   2011-01-01     single  2h93pZq0e7k5yf4dywlkpM         56  
0    2016-08-20      album  2h93pZq0e7k5yf4dywlkpM         81  
1    2012-01-01      album  2h93pZq0e7k5yf4dywlkpM         74  
2    2012-01-01      album  2h93pZq0e7k5yf4dywlkpM         33  
3    2012-01-01      album  2h93pZq0e7k5yf4dywlkpM         72  

[939 rows x 7 columns]

In [99]:
album_df.to_csv("album.csv", index = False)


## Populate the Track Table

- I definitely would use the sleep function on this one to make sure that I don't blow up the system
- Initially, you will be making 939 requests, which takes a bit, but didn't give me errors when I did it with the albums table
- I would recommend doing a request for 500 every 5 minutes
- You will have to use album table to get all tracks in each album
- also, you can just load in the album and artist table so you don't overdo it

## Script Notes

### Search Functions
Because more than one artist can appear when an artist is searched and
the search results aren't necessarily ranked, I have decided to find the search result with the
most amount of followers and return that artist as the "true artist" when determining ID's. This
assumes that the "true artist" will be the one with the most followers as it is the best metric
available. If this were a real music festival, we would have thse ID's on hand to be 100% sure
we are referencing the right artist. However, this lets me use spotify's search endpoint and save
some time in the process. 